src/Development/IDE/Core/Shake.hs

```diff
@@ -258,7 +258,22 @@ setValues :: IdeRule k v
           -> IO ()
 setValues state key file val = modifyVar_ state $ \vals -> do
     -- Force to make sure the old HashMap is not retained
-    evaluate $ HMap.insert (file, Key key) (fmap toDyn val) vals
+    let v = fmap toDyn val
+    _ <- evaluate v
+    let k = (file, Key key)
+    res <- evaluate $ HMap.insert k v vals
+
+    -- This odd construction is required to prevent a space leak.
+    -- We insert a fully evaluated value, but, somehow, the HashMap
+    -- seems to keep track of the old value. I don't know how.
```

Comment:
Are you sure that it is the old value that is being retained and not the old key or something else?

Reply:
I would imagine that if it were the old key, then the lookup and forcing that to Just/Nothing would be sufficient. But it's not. However, I only conclude its the value by process of elimination, so other ideas welcome.

Reply:
A reasonably reliable way to verify what is retained is to set a weak pointer on it.

Reply:
I put a weak ref on the value in the map at the start. I then did an insertion and the value was still available. I then did the evaluate trick on the value looked up from the Map, and the value was still available. I did a GC before both those things. I don't get how they remain available. My guess is if we solve the leak it would print DEREF VALUE STILL LIVES. If the leak never existed it would print nothing. But it prints both DEREF VALUE STILL LIVES and CONTINUE for each line. I put the code in a gist at https://gist.github.com/ndmitchell/7c62c03cc0966ebfa108e07e0549943a - hopefully I've done something wrong.

Reply:
I could look in gdb to find out what is retaining it if there is an easy to reproduce branch somewhere.

Reply:
I'm running the benchmark at https://gist.github.com/ndmitchell/11467985dbf1855e62035fa97248a585, on Windows GHC 8.6.5. That benchmark just loads the Shake repo and then does hover requests on a listMaybe value in a loop. Code compiled at `-O1`. You may wish to use Shake HEAD as otherwise there are other leaks that I've since eliminated. There is also a small leak with GetModifiedFiles I haven't tracked down, but this one is 10x that one.

Reply:
> I put a weak ref on the value in the map at the start. I then did an insertion and the value was still available. I then did the evaluate trick on the value looked up from the Map, and the value was still available. I did a GC before both those things. I don't get how they remain available. My guess is if we solve the leak it would print DEREF VALUE STILL LIVES. If the leak never existed it would print nothing. But it prints both DEREF VALUE STILL LIVES and CONTINUE for each line. I put the code in a gist at https://gist.github.com/ndmitchell/7c62c03cc0966ebfa108e07e0549943a - hopefully I've done something wrong.

This is easy to explain - the old value is still alive because the `state` var is still pointing at the old version of the hashmap

Reply:
This [gist](https://gist.github.com/pepeiborra/06ea0c5e4b4982bb1b0e0a9263f12634) shows that the values get released only after forcing the HashMap completely, and not in all cases, which means 3 things:

1. Weak ptrs work as expected,
2. The strict `HashMap` is not quite completely strict, so the fix in this PR is needed
3. Something else keeps some values alive, which is not surprising.

Example output:
```
DEREF VALUE STILL LIVES AFTER UPDATE
DEREF VALUE STILL LIVES AFTER EVALUATING THE MAP
DEREF VALUE STILL LIVES AFTER LOOKUP
DEREF VALUE STILL LIVES AFTER UPDATE
DEREF VALUE STILL LIVES AFTER EVALUATING THE MAP
DEREF VALUE STILL LIVES AFTER LOOKUP
DEREF VALUE STILL LIVES AFTER FORCING THE MAP
AN OLD VALUE GOT RELEASED
DEREF VALUE STILL LIVES AFTER UPDATE
DEREF VALUE STILL LIVES AFTER EVALUATING THE MAP
DEREF VALUE STILL LIVES AFTER LOOKUP
OLD VALUE IS GONE
AN OLD VALUE GOT RELEASED
OLD VALUE IS GONE
AN OLD VALUE GOT RELEASED
DEREF VALUE STILL LIVES AFTER UPDATE
DEREF VALUE STILL LIVES AFTER EVALUATING THE MAP
DEREF VALUE STILL LIVES AFTER LOOKUP
OLD VALUE IS GONE
AN OLD VALUE GOT RELEASED
DEREF VALUE STILL LIVES AFTER UPDATE
DEREF VALUE STILL LIVES AFTER EVALUATING THE MAP
DEREF VALUE STILL LIVES AFTER LOOKUP
[DEBUG] Cancelled request IdInt 39
```

Reply:
> because the state var is still pointing at the old version of the hashmap

Doh! Thanks.

Now we can prove that's the underlying cause, we can turn our attention to HashMap. Lots of things rely on it, so a space leak there must be wreaking havoc on loads of stuff.

Reply:
I've now reproduced this bug on top of unordered-containers alone. The key is that it only appears to trigger if you have hash collisions. That leads quickly to the question of why the hell we have so many hash collisions. I think the answer is that the Hashable of Key is dodgy, and we should be hashing in type information or all different mono-constructor's end up with a constant hash. Happily, I can fix both things, so will do so!